### PR TITLE
Fix for broken sorting of LUT file list

### LINKED ISSUE #11560: 3d LUT: Files in a LUT folder are listed in random order
After a LUT folder is selected, all contained LUT files can be scrolled through on the top right corner of the module. Yet the files appear to be listed in no determinable order. This behaviour has been around for years.

**Expected behavior**

The files should be listed alphabetically or in an order selected by the user. 

**Platform**

* darktable version : 3.9.0~git1228.cfb6a8cbf9-1
* OS                : Linux 5.13.0-40-generic #45-Ubuntu
* Linux - Distro    : Ubuntu 21.10

### PR DESCRIPTION
**Pull request by victoryforce**
Resolves #11560.

### CODE REVIEW

File: src/win/scandir.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    The original implementation of scandir has been made by Richard Salz.
+    The original author put this code in the public domain.
+
+    It has been modified by me to simplify slightly and increae readability.
+*/
+
+#pragma once
+
+int alphasort(const struct dirent**, const struct dirent**);
```
Comment (TurboGit) [inline src/win/scandir.h:30]: to follow dt convention should be named : `dt_scandir_alphasort`

File: src/win/scandir.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+int scandir(const char *directory_name,
```
Comment (TurboGit) [inline src/win/scandir.h:32]: `dt_scandir_scandir`

**Review by TurboGit — COMMENTED** [review]
I've made two minor comments, but more generally can't we use GLib API?

https://developer.gimp.org/api/2.0/glib/glib-File-Utilities.html

You could :
- Read all entries as done today and put them in a `GList`
- Sort the GList (see `g_list_sort`)

All this would not need a scandir implementation for Windows and using only standard API that dt is already using.

**Comment by TurboGit** [comment]
Hum... I thought there was some kind of scandir in Glib but I cannot see it now. Anyway, probably better for the scandir implementation to return a GList and void all the `malloc()` done yourself and the` struct dirent ***arraypointer` parameter.

**Comment by victoryforce** [comment]
> Hum... I thought there was some kind of scandir in Glib but I cannot see it now.

If it was, then of course there was no need to write a replacement function (even if only for mingw), since it's better to use the GLib API. But there is no such function in GLib.

> Anyway, probably better for the scandir implementation to return a GList and void all the `malloc()` done yourself and the` struct dirent ***arraypointer` parameter.

It will not be scandir, well known to everyone for decades (as BSD scandir), but another function that is similar to scandir, but for some reason has differences (like: why not reinvent the wheel?).  Personally, I've been using scandir function for probably thirty years, if not more. I don't see what we would achieve by implementing a very similar but incompatible function.

In my proposal, darktable will use the system function on almost all platforms where it can be built (with the exception of Windows, where our replacement will be used).

BTW, in the same lut3d.c we have another example of widespread convenient function that is missing from mingw so for Windows build it is simply added as source. This is getline and for Windows its source was added as win/getdelim.c by Peter Budai.

**Comment by TurboGit** [comment]
@victoryforce : Hum... how to say that... Well I did completely missed the goal of this PR in first read. Sorry!

I'll review again in the coming days.

File: src/iop/lut3d.c [hunk 3]
```diff
@@ -1504,29 +1501,25 @@ static void update_filepath_combobox(dt_iop_lut3d_gui_data_t *g, char *filepath,
     // new folder -> update the files list
     char *relativepath = g_path_get_dirname(filepath);
     char *folder = g_build_filename(lutfolder, relativepath, NULL);
-    struct dirent *dir;
-    DIR *d = opendir(folder);
-    if (d)
+
+    struct dirent **entries;
+    int numentries;
```
Comment (TurboGit) [inline src/iop/lut3d.c:1506]: const int numentries = scandir(folder, &entries, check_extension, alphasort);

File: src/iop/lut3d.c [hunk 4]
```diff
@@ -1504,29 +1501,25 @@ static void update_filepath_combobox(dt_iop_lut3d_gui_data_t *g, char *filepath,
...
-          filepath_set_unix_separator(ofilepath);
-          dt_bauhaus_combobox_add(g->filepath, ofilepath);
-          g_free(ofilepath);
-        }
-      }
-      dt_bauhaus_widget_t *w = DT_BAUHAUS_WIDGET(g->filepath);
-      dt_bauhaus_combobox_data_t *combo_data = &w->data.combobox;
-      g_ptr_array_sort(combo_data->entries, array_str_cmp);
-      closedir(d);
+      char *file = entries[i]->d_name;
```
Comment (TurboGit) [inline src/iop/lut3d.c:1512]: const

File: src/win/scandir.c [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+{
+  DIR *directory;
+  struct dirent **array;
+  struct dirent *entry;
+  struct dirent *copy;
+  int allocated = INITIAL_ALLOCATION;
+  int counter = 0;
+
+  /* Get initial list space and open directory.  */
+  directory = opendir(directory_name);
```
Comment (TurboGit) [inline src/win/scandir.c:57]: DIR *directory = ...;

File: src/win/scandir.c [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+/* Initial guess at directory allocated.  */
+#define INITIAL_ALLOCATION 20
+
+int scandir(const char *directory_name,
+            struct dirent ***array_pointer,
+            int (*select_function) (const struct dirent *),
+            int (*compare_function) (const struct dirent**, const struct dirent**))
+{
+  DIR *directory;
```
Comment (TurboGit) [inline src/win/scandir.c:49]: remove, and see below

File: src/win/scandir.c [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+       *                giving the name of the directory entry. This name is
+       *                null terminated and may have at most MAXNAMLEN chars.
+       * So our malloc length may need to be increased accordingly.
+       *        sizeof(entry->d_name): space (possibly minimal) in struct.
+       *         strlen(entry->d_name): actual length of the entry. 
+       *
+       *                        John Kavadias <john_kavadias@hotmail.com>
+       *                        David Lee <t.d.lee@durham.ac.uk>
+       */
+      int namelength = strlen(entry->d_name) + 1;        /* length with NULL */
```
Comment (TurboGit) [inline src/win/scandir.c:86]: const int

File: src/win/scandir.c [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+      if(counter + 1 == allocated)
+      {
+        allocated <<= 1;
+        array = (struct dirent **)
+          realloc((char *) array, allocated * sizeof(struct dirent *));
+        if(array == NULL)
+        {
+          closedir(directory);
+          free(array);
```
Comment (TurboGit) [inline src/win/scandir.c:116]: in this path, array is null => not needed and will crash. Only `g_free()` cope with NULL pointer.

File: src/win/scandir.c [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+       *        sizeof(entry->d_name): space (possibly minimal) in struct.
+       *         strlen(entry->d_name): actual length of the entry. 
+       *
+       *                        John Kavadias <john_kavadias@hotmail.com>
+       *                        David Lee <t.d.lee@durham.ac.uk>
+       */
+      int namelength = strlen(entry->d_name) + 1;        /* length with NULL */
+      int extra = 0;
+
+      if(sizeof(entry->d_name) <= namelength)
```
Comment (TurboGit) [inline src/win/scandir.c:89]: Maybe only `<` because if `=` the computation below will set extra to 0 anyway.

**Review by TurboGit — CHANGES_REQUESTED** [review]
Not tested yet, some comments after review.

File: src/win/scandir.c (hunk 8 above)
Comment (parafin) [reply src/win/scandir.c:116]: free shouldn't crash on NULL, but anyway it doesn't make sense to call it here at all

Comment (TurboGit) [reply src/win/scandir.c:116]: Indeed, wrong recollection I thought it was a diff between g_free/free but it is not. My confusion comes from the fact that most `g_...()` routines working on strings can take a NULL pointer whereas the equivalent std function do not.

**Review by TurboGit — APPROVED** [review]
All good to me now, thanks!

**Comment by victoryforce** [comment]
Thanks for reviewing!

In addition, I want to record here some facts that may be useful to other developers. Of course, the issue comments aren't the best place for this, but I'll just write it down here for now, maybe later such things can be added to some page on the dev wiki.

### Should we check the parameter of the free() function for NULL?

Short answer: it's not worth it for modern platforms.

The correct processing of such a parameter has been guaranteed in the C standards for a long time (at least since C99, which in turn simply recorded long-existing implementations as a standard).

### Are there platforms that do not adhere to this standard and have an unsafe implementation of the free() function?

Somewhat bad news: Yes, they (reportedly) are. The good news: These plaftorms are completely obsolete and are only of historical interest. free(NULL) is reported to cause problems only on PalmOS, NetWare, and extremely old flavor of Unix known as 3BSD.

### Is it good practice to not bother checking for NULL before calling free() ?

Checking only adds unnecessary code to read and execute and is thus not a good practice. In fact, the same check will be repeated immediately after the function is called. One might say that such a check before free() can avoid an unnecessary function call. But the common case is that things don't fail and you will normally end up with allocated memory instead of NULL pointers. If you add extra checks, this means that you'll negatively (even if only minimally) impact the performance of your common cases just to avoid an extra function call in what's will be failure case (when performance is no longer important at all).

**Pull request merged**